fdbserver/GetEncryptCipherKeys.actor.cpp

```diff
@@ -0,0 +1,256 @@
+/*
+ * GetCipherKeys.actor.cpp
+ *
+ * This source file is part of the FoundationDB open source project
+ *
+ * Copyright 2013-2022 Apple Inc. and the FoundationDB project authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "fdbserver/GetEncryptCipherKeys.h"
+
+#include <boost/functional/hash.hpp>
+
+namespace {
+
+Optional<UID> getEncryptKeyProxyId(const Reference<AsyncVar<ServerDBInfo> const>& db) {
+	return db->get().encryptKeyProxy.map<UID>([](EncryptKeyProxyInterface proxy) { return proxy.id(); });
+}
+
+ACTOR Future<Void> onEncryptKeyProxyChange(Reference<AsyncVar<ServerDBInfo> const> db) {
+	state Optional<UID> previousProxyId = getEncryptKeyProxyId(db);
+	state Optional<UID> currentProxyId;
+	loop {
+		wait(db->onChange());
+		currentProxyId = getEncryptKeyProxyId(db);
+		if (currentProxyId != previousProxyId) {
+			break;
+		}
+	}
+	TraceEvent("GetCipherKeys_EncryptKeyProxyChanged")
+	    .detail("PreviousProxyId", previousProxyId.orDefault(UID()))
+	    .detail("CurrentProxyId", currentProxyId.orDefault(UID()));
+	return Void();
+}
+
+ACTOR Future<EKPGetLatestBaseCipherKeysReply> getUncachedLatestEncryptCipherKeys(
+    Reference<AsyncVar<ServerDBInfo> const> db,
+    EKPGetLatestBaseCipherKeysRequest request) {
+	Optional<EncryptKeyProxyInterface> proxy = db->get().encryptKeyProxy;
+	if (!proxy.present()) {
+		// Wait for onEncryptKeyProxyChange.
+		TraceEvent("GetLatestCipherKeys_EncryptKeyProxyNotPresent");
+		return Never();
+	}
+	request.reply.reset();
+	try {
+		EKPGetLatestBaseCipherKeysReply reply = wait(proxy.get().getLatestBaseCipherKeys.getReply(request));
+		if (reply.error.present()) {
+			TraceEvent("GetLatestCipherKeys_RequestFailed").error(reply.error.get());
+			throw reply.error.get();
```

Comment:
do we want to filter the error thrown upstream OR any error needs to be propogated?

Reply:
Do you have any handling in mind? It seems to me with either encryption error or unexpected error of some other type, the handling is just to propogate.

Reply:
I think converting to encrypt_keys_fetch_failed error might be better. will update.